fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -27,197 +27,495 @@
+		auto const limitingCostForTag = limitingCost.get() * quotaRatio;
+		return limitingCostForTag / averageTransactionCost.get();
+	}
+
+	// Return the limiting transaction rate, aggregated across all storage servers
+	Optional<double> getLimitingTps(TransactionTag tag, OpType opType) const {
+		Optional<double> result;
+		for (const auto& [id, _] : throttlingRatios) {
+			auto const targetTpsForSS = getLimitingTps(id, tag, opType);
+			if (result.present() && targetTpsForSS.present()) {
+				result = std::min(result.get(), targetTpsForSS.get());
+			} else if (targetTpsForSS.present()) {
+				result = targetTpsForSS;
 			}
 		}
-	};
+		return result;
+	}
 
-	Database db;
-	UID id;
-	std::map<TransactionTag, QuotaAndCounters> trackedTags;
-	uint64_t throttledTagChangeId{ 0 };
-	Future<Void> traceActor;
+	Optional<double> getDesiredTps(TransactionTag tag, OpType opType, double averageTransactionCost) const {
+		auto const quota = getQuota(tag, opType, LimitType::TOTAL);
+		if (!quota.present()) {
+			return {};
+		}
+		return quota.get() / averageTransactionCost;
+	}
 
-	ACTOR static Future<Void> tracer(GlobalTagThrottlerImpl const* self) {
-		loop {
-			for (const auto& [tag, quotaAndCounters] : self->trackedTags) {
-				TraceEvent te("GlobalTagThrottling");
-				te.detail("Tag", tag);
-				quotaAndCounters.processTraceEvent(te);
+	Optional<double> getDesiredTps(TransactionTag tag,
+	                               double averageTransactionReadCost,
+	                               double averageTransactionWriteCost) const {
+		auto const readDesiredTps = getDesiredTps(tag, OpType::READ, averageTransactionReadCost);
+		auto const writeDesiredTps = getDesiredTps(tag, OpType::WRITE, averageTransactionWriteCost);
+		if (readDesiredTps.present() && writeDesiredTps.present()) {
+			return std::min(readDesiredTps.get(), writeDesiredTps.get());
+		} else if (readDesiredTps.present()) {
+			return readDesiredTps;
+		} else {
+			return writeDesiredTps;
+		}
+	}
+
+	Optional<double> getReservedTps(TransactionTag tag, OpType opType, double averageTransactionCost) const {
+		auto const reservedCost = getQuota(tag, opType, LimitType::RESERVED);
+		if (!reservedCost.present()) {
+			return {};
+		} else {
+			return reservedCost.get() / averageTransactionCost;
+		}
+	}
+
+	Optional<double> getReservedTps(TransactionTag tag,
+	                                double averageTransactionReadCost,
+	                                double averageTransactionWriteCost) const {
+		auto const readReservedTps = getReservedTps(tag, OpType::READ, averageTransactionReadCost);
+		auto const writeReservedTps = getReservedTps(tag, OpType::WRITE, averageTransactionWriteCost);
+		if (readReservedTps.present() && writeReservedTps.present()) {
+			return std::max(readReservedTps.get(), writeReservedTps.get());
+		} else if (readReservedTps.present()) {
+			return readReservedTps;
+		} else {
+			return writeReservedTps;
+		}
+	}
```

Comment:
For the `getReservedTps` and `getDesiredTps` share the same logic except for `LIMIT_TYPE`, IMO it's better the same function like `getTps(tag, readCost, writeCost, limitType)` to handle both.